compiler/rustc_trait_selection/src/error_reporting/infer/need_type_info.rs

```diff
@@ -613,6 +677,124 @@ impl<'a, 'tcx> TypeErrCtxt<'a, 'tcx> {
+                // `From` impls that correspond to the self type as well, so as to avoid showing
+                // multiple conversion options.
+                Some(assoc_ty)
+            } else {
+                None
+            };
+            let assocs = tcx.associated_items(impl_def_id);
+
+            if tcx.is_diagnostic_item(sym::blanket_into_impl, impl_def_id)
+                && let Some(did) = tcx.get_diagnostic_item(sym::From)
+            {
+                let mut found = false;
+                tcx.for_each_impl(did, |impl_def_id| {
+                    // We had an `<A as Into<B>::into` and we've hit the blanket
+                    // impl for `From<A>`. So we try and look for the right `From`
+                    // impls that *would* apply. We *could* do this in a generalized
+                    // version by evaluating the `where` clauses, but that would be
+                    // way too involved to implement. Instead we special case the
+                    // arguably most common case of `expr.into()`.
+                    let Some(header) = tcx.impl_trait_header(impl_def_id) else {
+                        return;
+                    };
+                    let target = header.trait_ref.skip_binder().args.type_at(0);
+                    if filter.is_some() && filter != Some(target) {
+                        return;
+                    };
+                    let target = header.trait_ref.skip_binder().args.type_at(0);
+                    let ty = header.trait_ref.skip_binder().args.type_at(1);
+                    if ty == self_ty {
+                        if target_type {
+                            let mut ty_str = format!("{target}");
+                            if &ty_str == "_" {
+                                ty_str = "/* Type */".to_string();
+                            }
+                            paths.push(ty_str);
+                        } else {
+                            paths.push(format!("<{self_ty} as Into<{target}>>::into"));
+                        }
+                        found = true;
+                    }
+                });
+                if found {
+                    return;
+                }
+            }
+
+            // We're at the `impl` level, but we want to get the same method we
+            // called *on this `impl`*, in order to get the right DefId and args.
+            let Some(assoc) = assocs.filter_by_name_unhygienic(name).next() else {
+                // The method isn't in this `impl`? Not useful to us then.
+                return;
+            };
+            let Some(trait_assoc_item) = assoc.trait_item_def_id else {
+                return;
+            };
+            let trait_assoc_substs =
+                impl_trait_ref
+                    .args
+                    .extend_to(tcx, trait_assoc_item, |def, _| self.var_for_def(DUMMY_SP, def));
+            let identity_method = args.rebase_onto(tcx, def_id, trait_assoc_substs);
```

Comment:
What is happening here? You're rebasing args *onto* method args? I don't think this is right.